### PR TITLE
Add -lightwalletd experimental option

### PR DESCRIPTION
**Pull request by oxarbitrage**
Similar to `-insightexplorer` but loading less indexes.

After testing and code review this should be able to close https://github.com/zcash/zcash/issues/4326

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -70,7 +70,7 @@ int nScriptCheckThreads = 0;
 bool fImporting = false;
 std::atomic_bool fReindex(false);
 bool fTxIndex = false;
-bool fAddressIndex = false;     // insightexplorer
+bool fAddressIndex = false;     // insightexplorer && lightwalletd
```
Comment (str4d) [inline src/main.cpp]:
```suggestion
bool fAddressIndex = false;     // insightexplorer || lightwalletd
```

File: src/main.cpp [hunk 2]
```diff
@@ -4477,6 +4477,15 @@ bool static LoadBlockIndexDB()
...
     fTimestampIndex = fInsightExplorer;
 
+    bool fLightWalletd = false;
+    pblocktree->ReadFlag("lightwalletd", fLightWalletd);
+    LogPrintf("%s: light wallet daemon %s\n", __func__, fLightWalletd ? "enabled" : "disabled");
+    if (fLightWalletd) {
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
+    }
```
Comment (str4d) [inline src/main.cpp]: If both `-insightexplorer` and `-lightwalletd` are set, this causes `-insightexplorer` to be ignored. We should either make them mutually exclusive (only one can be set, enforced by `init.cpp`), or handle the case here where both are set. I'm weakly leaning towards making them mutually exclusive.

File: src/main.cpp [hunk 3]
```diff
@@ -4821,6 +4830,14 @@ bool InitBlockIndex(const CChainParams& chainparams)
     fSpentIndex = fExperimentalInsightExplorer;
     fTimestampIndex = fExperimentalInsightExplorer;
 
+    // Use the provided setting for -lightwalletd in the new database
+    pblocktree->WriteFlag("lightwalletd", fExperimentalLightWalletd);
+    if (fExperimentalLightWalletd) {
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
+    }
```
Comment (str4d) [inline src/main.cpp]: Ditto.

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -408,8 +408,8 @@ UniValue getrawmempool(const UniValue& params, bool fHelp)
 UniValue getblockdeltas(const UniValue& params, bool fHelp)
 {
     std::string disabledMsg = "";
-    if (!fExperimentalInsightExplorer) {
-        disabledMsg = experimentalDisabledHelpMsg("getblockdeltas", "insightexplorer");
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
Comment (str4d) [inline src/rpc/blockchain.cpp]:
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -462,7 +462,7 @@ UniValue getblockdeltas(const UniValue& params, bool fHelp)
             + HelpExampleRpc("getblockdeltas", "\"00227e566682aebd6a7a5b772c96d7a999cadaebeaf1ce96f4191a3aad58b00b\"")
         );
 
-    if (!fExperimentalInsightExplorer) {
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
Comment (str4d) [inline src/rpc/blockchain.cpp]:
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -491,8 +491,8 @@ UniValue getblockdeltas(const UniValue& params, bool fHelp)
 UniValue getblockhashes(const UniValue& params, bool fHelp)
 {
     std::string disabledMsg = "";
-    if (!fExperimentalInsightExplorer) {
-        disabledMsg = experimentalDisabledHelpMsg("getblockhashes", "insightexplorer");
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
Comment (str4d) [inline src/rpc/blockchain.cpp]:
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```

File: src/rpc/blockchain.cpp [hunk 7]
```diff
@@ -525,7 +525,7 @@ UniValue getblockhashes(const UniValue& params, bool fHelp)
             + HelpExampleCli("getblockhashes", "1558141697 1558141576 '{\"noOrphans\":false, \"logicalTimes\":true}'")
             );
 
-    if (!fExperimentalInsightExplorer) {
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
Comment (str4d) [inline src/rpc/blockchain.cpp]:
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```
and ditto below.

File: src/rpc/server.cpp [hunk 8]
```diff
@@ -467,14 +467,43 @@ std::string HelpExampleRpc(const std::string& methodname, const std::string& arg
...
 
-string experimentalDisabledHelpMsg(const string& rpc, const string& enableArg)
-{
-    return "\nWARNING: " + rpc + " is disabled.\n"
-        "To enable it, restart zcashd with the -experimentalfeatures and\n"
-        "-" + enableArg + " commandline options, or add these two lines\n"
-        "to the zcash.conf file:\n\n"
-        "experimentalfeatures=1\n"
-        + enableArg + "=1\n";
+string experimentalDisabledHelpMsg(const string& rpc, const vector<string>& enableArgs)
```
Comment (str4d) [inline src/rpc/server.cpp]: While we are changing this line, make these:
```suggestion
std::string experimentalDisabledHelpMsg(const std::string& rpc, const std::vector<string>& enableArgs)
```

**Comment by oxarbitrage** [comment]
Thanks for the fast and positive review @str4d .

I addressed all the review comments in https://github.com/zcash/zcash/pull/4402/commits/0a194bd6f6a273645ed9a84896dfd8fd65dc8e36 , we can force push after a new review of the changes.

File: src/main.cpp (hunk 2 above)
Comment (daira) [reply src/main.cpp]: Can't this be solved by just setting `fAddressIndex = true;` here? It's not logically inconsistent to want to support both the Insight Explorer and lightwalletd from the same zcashd.

File: qa/rpc-tests/addressindex.py [hunk 9]
```diff
@@ -291,6 +307,15 @@ def check_balance(node_index, address, expected_balance, expected_received=None)
...
             assert_equal(deltas[i]['txid'],     expected_deltas[i]['txid'])
 
+        # test getaddressdeltas in the lightwalletd node
+        deltas = self.nodes[3].getaddressdeltas({'addresses': [addr1]})
+        assert_equal(len(deltas), len(expected_deltas))
+        for i in range(len(deltas)):
+            assert_equal(deltas[i]['address'],  addr1)
+            assert_equal(deltas[i]['height'],   expected_deltas[i]['height'])
+            assert_equal(deltas[i]['satoshis'], expected_deltas[i]['satoshis'])
+            assert_equal(deltas[i]['txid'],     expected_deltas[i]['txid'])
```
Comment (daira) [inline qa/rpc-tests/addressindex.py]: To avoid code duplication, I suggest:
```
        # test getaddressdeltas
        for node in (1, 3):
            deltas = self.nodes[node].getaddressdeltas({'addresses': [addr1]})
            assert_equal(len(deltas), len(expected_deltas))
            for i in range(len(deltas)):
                assert_equal(deltas[i]['address'],  addr1)
                assert_equal(deltas[i]['height'],   expected_deltas[i]['height'])
                assert_equal(deltas[i]['satoshis'], expected_deltas[i]['satoshis'])
                assert_equal(deltas[i]['txid'],     expected_deltas[i]['txid'])
```

File: qa/rpc-tests/addressindex.py [hunk 10]
```diff
@@ -255,6 +266,11 @@ def check_balance(node_index, address, expected_balance, expected_received=None)
         change = change_vout[0]['scriptPubKey']['addresses'][0]
         bal = self.nodes[2].getaddressbalance(change)
         assert(bal['received'] > 0)
+
+        # test getaddressbalance for the lightwalletd node
+        bal = self.nodes[3].getaddressbalance(change)
+        assert(bal['received'] > 0)
```
Comment (daira) [inline qa/rpc-tests/addressindex.py]:
```suggestion
        # test getaddressbalance
        for node in (2, 3):
            bal = self.nodes[3].getaddressbalance(change)
            assert(bal['received'] > 0)
```

File: src/main.cpp [hunk 11]
```diff
@@ -4468,14 +4468,24 @@ bool static LoadBlockIndexDB()
...
+    LogPrintf("%s: light wallet daemon %s\n", __func__, fLightWalletd ? "enabled" : "disabled");
+    if (fInsightExplorer) {
+        fAddressIndex = true;
+        fSpentIndex = true;
+        fTimestampIndex = true;
+    }
+    else if (fLightWalletd) {
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
```
Comment (daira) [inline src/main.cpp]:
```suggestion
```

File: src/init.cpp [hunk 12]
```diff
@@ -1418,6 +1418,20 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
+                    strLoadError = _("You need to rebuild the database using -reindex to change -lightwalletd");
+                    break;
+                }
+
+                // -insightexplorer and -lightWalletd are mutually exclusive
+                if (fInsightExplorerPreviouslySet && fLightWalletdPreviouslySet) {
+                    strLoadError = _("Not possible to enable -insightexplorer and -lightwalletd at the same time");
+                    break;
+                }
+
```
Comment (daira) [inline src/init.cpp]:
```suggestion
```
(requires my other suggested changes below)

File: src/main.cpp [hunk 13]
```diff
@@ -4815,11 +4825,19 @@ bool InitBlockIndex(const CChainParams& chainparams)
...
+    pblocktree->WriteFlag("lightwalletd", fExperimentalLightWalletd);
+    if (fExperimentalInsightExplorer) {
+        fAddressIndex = true;
+        fSpentIndex = true;
+        fTimestampIndex = true;
+    }
+    else if (fExperimentalLightWalletd) {
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
```
Comment (daira) [inline src/main.cpp]:
```suggestion
```

File: src/init.cpp (hunk 12 above)
Comment (daira) [reply src/init.cpp]: If this is not taken, then `s/-lightWalletd/-lightwalletd/` in the comment.

File: src/rpc/server.cpp [hunk 14]
```diff
@@ -467,14 +467,43 @@ std::string HelpExampleRpc(const std::string& methodname, const std::string& arg
...
+            config += newline() + "or:" + newline(2);
+            config += "experimentalfeatures=1" + newline();
+            config += enableArgs.at(i) + "=1" + newline();
+        }
+    }
+    return "\nWARNING: " + rpc + " is disabled." + newline() +
+        "To enable it, restart zcashd with the following command line options:" + newline()
+        + cmd + newline(2) +
+        "Alternatively add these two lines to the zcash.conf file:" + newline(2)
+        + config;
```
Comment (daira) [inline src/rpc/server.cpp]:
```suggestion
    for (size_t i = 0; i < size; ++i)
    {
        if (size == 1 || i == 0)
        {
            cmd += "-experimentalfeatures and -" + enableArgs.at(i);

            config += "experimentalfeatures=1\n";
            config += enableArgs.at(i) + "=1\n";
        }
        else {
            cmd += " or:\n-experimentalfeatures and -" + enableArgs.at(i);

            config += "\nor:\n\n";
            config += "experimentalfeatures=1\n";
            config += enableArgs.at(i) + "=1\n";
        }
    }
    return "\nWARNING: " + rpc + " is disabled.\n" +
        "To enable it, restart zcashd with the following command line options:\n"
        + cmd + "\n\n" +
        "Alternatively add these two lines to the zcash.conf file:\n\n"
        + config;
```

File: src/main.cpp (hunk 2 above)
Comment (oxarbitrage) [reply src/main.cpp]: Went with @daira suggestion at https://github.com/zcash/zcash/pull/4402/commits/7aed10f195c961408e6c4bbbae2595c2c17e5a89

File: src/rpc/server.cpp (hunk 14 above)
Comment (oxarbitrage) [reply src/rpc/server.cpp]: Yes, i had it like this before, then added the lambda to clarify a bit the code but actually was not a very good idea. Removed at https://github.com/zcash/zcash/pull/4402/commits/5d3ff01d6177e5fe3974b4483920d794f7e37f12

File: src/rpc/misc.cpp [hunk 15]
```diff
@@ -1063,8 +1063,8 @@ UniValue getaddresstxids(const UniValue& params, bool fHelp)
 UniValue getspentinfo(const UniValue& params, bool fHelp)
 {
     std::string disabledMsg = "";
-    if (!fExperimentalInsightExplorer) {
-        disabledMsg = experimentalDisabledHelpMsg("getspentinfo", "insightexplorer");
+    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
+        disabledMsg = experimentalDisabledHelpMsg("getspentinfo", {"insightexplorer", "lightwalletd"});
```
Comment (str4d) [inline src/rpc/misc.cpp]: Revert these changes; `getspentinfo` cannot be called with `-lightwalletd` because this RPC requires `fSpentIndex` which `-lightwalletd` does not enable.

File: src/rpc/misc.cpp [hunk 16]
```diff
@@ -1087,7 +1087,7 @@ UniValue getspentinfo(const UniValue& params, bool fHelp)
             + HelpExampleRpc("getspentinfo", "{\"txid\": \"33990288fb116981260be1de10b8c764f997674545ab14f9240f00346333b780\", \"index\": 4}")
         );
 
-    if (!fExperimentalInsightExplorer) {
+    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```
Comment (str4d) [inline src/rpc/misc.cpp]: Ditto.

File: src/rpc/misc.cpp (hunk 15 above)
Comment (oxarbitrage) [reply src/rpc/misc.cpp]: Right, fixed at https://github.com/zcash/zcash/pull/4402/commits/f65acd0aeea44cc4d3ab058ed8b870fb3ba5b595

File: qa/rpc-tests/addressindex.py [hunk 17]
```diff
@@ -255,6 +266,12 @@ def check_balance(node_index, address, expected_balance, expected_received=None)
         change = change_vout[0]['scriptPubKey']['addresses'][0]
         bal = self.nodes[2].getaddressbalance(change)
         assert(bal['received'] > 0)
```
Comment (str4d) [inline qa/rpc-tests/addressindex.py]: Remove these two lines (they are checked immediately below).

**Comment by oxarbitrage** [comment]
Removed duplicated lines from test, squashed everything with the first commit by `git revise` and force pushed.

**Comment by LarryRuane** [comment]
> ... (without rebasing on master, so GitHub's force-push UI shows no change) ...

@str4d, what is the GitHub force-push UI?

**Comment by str4d** [comment]
@LarryRuane when you force-push to a branch that has a PR, GitHub adds a note to the PR's timeline. Clicking on the words "force-pushed" shows the changes across the force-push. [In this case](https://github.com/zcash/zcash/compare/f65acd0aeea44cc4d3ab058ed8b870fb3ba5b595..4d0b20882cb4ff728901018828526584131709d7), the only changes were the ones that addressed my comment.

**Comment by str4d** [comment]
Nice work!

@zkbot r+

**Comment by oxarbitrage** [comment]
Failed by pyflakes: `qa/rpc-tests/addressindex.py:19: 'test_framework.util.start_node' imported but unused`.

Fixed and force pushed, please retry.

**Comment by str4d** [comment]
Confirmed that the force-push [only adjusted the test imports](https://github.com/zcash/zcash/compare/4d0b20882cb4ff728901018828526584131709d7..a43f5daff8819520da32ea7569033625b8ca656f).

@zkbot r+

**Comment by str4d** [comment]
An unexpected transient failure; opened #4438.

@zkbot retry

**Comment by str4d** [comment]
More transient failures...

@zkbot retry
